src/sage/schemes/hyperelliptic_curves/jacobian_endomorphisms.py

```diff
@@ -0,0 +1,465 @@
+    return (False, False)
+
+
+def is_geom_trivial_when_field(C, bad_primes, B=200):
+    """
+    Codes up Lombardo's Algorithm 4.15
+
+    This assumes that the Jacobian of the curve C is absolutely simple.
+
+    INPUT:
+
+    - ``C`` -- the hyperelliptic curve.
+
+    - ``bad_primes`` -- the list of odd primes of bad reduction.
+
+    - ``B`` -- the bound which appears in the statement of the algorithm
+    from Lombardo's paper.
+
+    OUTPUT:
+
+        (bool): boolean indicating whether or not the absolute endomorphism
+        algebra is a field
+
+    EXAMPLES::
+
+    This is LMFDB curve 461.a.461.2
+
+        sage: from sage.schemes.hyperelliptic_curves.jacobian_endomorphisms import is_geom_trivial_when_field
+        sage: R.<x> = QQ[]
+        sage: f = 4*x^5 - 4*x^4 - 156*x^3 + 40*x^2 + 1088*x - 1223
+        sage: C = HyperellipticCurve(f)
+        sage: is_geom_trivial_when_field(C,[461])
+        True
+
+    This is LMFDB curve 4489.a.4489.1
+
+        sage: f = x^6 + 4*x^5 + 2*x^4 + 2*x^3 + x^2 - 2*x + 1
+        sage: C = HyperellipticCurve(f)
+        sage: is_geom_trivial_when_field(C,[67])
+        False
+    """
+
+    running_gcd = 0
+    R = PolynomialRing(ZZ,2,"xv")
+    x,v = R.gens()
+    T = PolynomialRing(QQ,'v')
+    g = v - x**4
+
+    for p in prime_range(3,B):
+        if p not in bad_primes:
+            Cp = C.change_ring(FiniteField(p))
+            fp = Cp.frobenius_polynomial()
+            if satisfies_coefficient_condition(fp, p):
+                # This defines the polynomial f_v**[4] from the paper
+                fp4 = T(R(fp).resultant(g))
+                if fp4.is_irreducible():
+                    running_gcd = gcd(running_gcd, NumberField(fp,'a').discriminant())
+                    if running_gcd <= 24:
+                        return True
+    return False
+
+
+class EndomorphismRing(SageObject):
+    r"""
+    The ring of endomorphisms of this Jacobian which are defined over the
+    base field.
+
+    EXAMPLES::
+
+        sage: R.<x> = QQ[]
+        sage: f = x**5 + 17
+        sage: C = HyperellipticCurve(f)
+        sage: J = C.jacobian()
+        sage: E = J.endomorphism_ring()
+        sage: E
+        Endomorphism ring of Jacobian of Hyperelliptic Curve over Rational Field defined by y^2 = x^5 + 17
+    """
+
+    def __init__(self, A):
+        r"""
+        See ``EndomorphismRing`` for documentation.
+
+        EXAMPLES::
+
+            sage: R.<x> = QQ[]
+            sage: f = x^6 - 2*x^4 + 6*x^3 + x + 57
+            sage: C = HyperellipticCurve(f)
+            sage: J = C.jacobian()
+            sage: E = J.endomorphism_ring()
+            sage: E
+            Endomorphism ring of Jacobian of Hyperelliptic Curve over Rational Field defined by y^2 = x^6 - 2*x^4 + 6*x^3 + x + 57
+        """
+        self._A = A
+        self._have_established_absolutely_field = False
+        self._have_established_absolutely_trivial = False
+
```

Comment:
Using `lazy_attribute` https://doc.sagemath.org/html/en/reference/misc/sage/misc/lazy_attribute.html would make the code simpler

Reply:
Not sure how to do this ... have sent an email for clarification and my concerns

Reply:
As explained on trac, this is no longer going to be done with `lazy_attribute`, but a proof flag has been implemented, which raises a `NotImplementedError` if the methods return `False` and the user insists on `proof=True` (since lower bounds have not been implemented here).